prov/verbs/src/verbs_cq.c

```diff
@@ -230,47 +230,184 @@ static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
+	entry = slist_remove_first_match(&cq->ep_list,
+					 fi_ibv_match_ep_id,
+					 (void *)wc->wr_id);
 	if (!entry) {
 		VERBS_WARN(FI_LOG_CQ, "No matching EP for :"
 			   "given signaled send completion\n");
 		return -FI_EOTHER;
 	}
 	epe = container_of(entry, struct fi_ibv_msg_epe, entry);
 	ofi_atomic_sub32(&epe->ep->unsignaled_send_cnt,
-			VERBS_SEND_SIGNAL_THRESH(epe->ep));
+			 VERBS_SEND_SIGNAL_THRESH(epe->ep));
 	ofi_atomic_dec32(&epe->ep->comp_pending);
 	util_buf_release(cq->epe_pool, epe);
 
-	return 0;
+	return FI_SUCCESS;
+}
+
+static inline int fi_ibv_poll_outstanding_cq(struct fi_ibv_msg_ep *ep,
+					     struct fi_ibv_cq *cq,
+					     struct dlist_entry *out_list)
+{
+	struct fi_ibv_wre *wre;
+	struct fi_ibv_wce *wce;
+	struct fi_ibv_msg_ep *tmp_ep;
+	struct ibv_wc wc;
+	ssize_t ret;
+
+	dlist_foreach_container(out_list, struct fi_ibv_wre,
+				wre, entry) {
+		ret = 1;
+		goto done;
+	}
+
+	ret = ibv_poll_cq(cq->cq, 1, &wc);
+	if (ret <= 0)
+		return ret;
+
+	if ((wc.opcode == IBV_WC_RECV) ||
+	    (wc.opcode == IBV_WC_RECV_RDMA_WITH_IMM) ||
+	    ((wc.wr_id & cq->wr_id_mask) != cq->send_signal_wr_id))
+		goto done;
+
+	ret = fi_ibv_handle_internal_signal_wc(cq, &wc);
+	return ret ? ret : 1;
+done:
+	wre = (struct fi_ibv_wre *)(uintptr_t)wc.wr_id;
+	dlist_remove(&wre->entry);
+	if (wre->ep.ep) {
+		if ((wre->ep.ep != ep) &&
+		    (wc.status != IBV_WC_WR_FLUSH_ERR)) {
+			wce = util_buf_alloc(cq->wce_pool);
+			if (!wce) {
+				dlist_insert_tail(&wre->entry,
+						  out_list);
+				return -FI_EAGAIN;
+			}
+			memset(wce, 0, sizeof(*wce));
+			wc.wr_id = (uintptr_t)wre->context;
+			wce->wc = wc;
+			slist_insert_tail(&wce->entry, &cq->wcq);
+		}
+	} else {
+		/* if `ep == NULL` it means that WRE is already released */
+		return ret;
+	}
+	tmp_ep = wre->ep.ep;
+	wre->ep.ep = NULL;
+	util_buf_release(tmp_ep->wre_pool, wre);
+	return ret;
+}
+
+void fi_ibv_cleanup_cq(struct fi_ibv_msg_ep *ep)
+{
+	DEFINE_LIST(wre_out_list);
+	int ret;
+	struct fi_ibv_wre *wre;
+	struct dlist_entry *tmp;
+
+	fastlock_acquire(&ep->rcq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->rcq,
+						 &wre_out_list);
+	} while (ret > 0 || ret == -FI_EAGAIN);
+	/* Handle WRs for which there were no appropriate WCs */
+	dlist_foreach_container_safe(&ep->wre_list, struct fi_ibv_wre,
+				     wre, entry, tmp) {
+		if ((ep == wre->ep.ep) &&
+		    (wre->wr.ibv_wr_type == IBV_RECV_WR)) {
+			dlist_remove(&wre->entry);
+			wre->ep.ep = NULL;
+			util_buf_release(ep->wre_pool, wre);
+		}
+	}
+	fastlock_release(&ep->rcq->lock);
+
+	fastlock_acquire(&ep->scq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->scq,
+						 &wre_out_list);
+	} while (ret > 0 || ret == -FI_EAGAIN);
+	dlist_foreach_container_safe(&ep->wre_list, struct fi_ibv_wre,
+				     wre, entry, tmp) {
+		if ((ep == wre->ep.ep) &&
+		    (wre->wr.ibv_wr_type == IBV_SEND_WR)) {
+			dlist_remove(&wre->entry);
+			wre->ep.ep = NULL;
+			util_buf_release(ep->wre_pool, wre);
```

Comment:
It seems one of these dlist travesal should be enough (after polling scq and rcq).

Reply:
Not sure, because we can get unpredictable event. Let's assume the following case:
1. We posted 200 recv and 200 send WRs for some EP.
2. During polling scq, rcq. We read 100 recv and 100 send WCs for this EP.
3. There is possibility that some WRs is completed in another thread (that polls send and recv CQ) with bad/suc status (for example, remote side closes EP and we can get WCs for the remaining WRs)